text/0088-contract-level-messages.md

```diff
@@ -0,0 +1,237 @@
+
+```rust
+pub enum Key {
+    ...
+    /// A `Key` under which a message topic or message checksum is stored.
+    Message(MessageAddr),
+}
+```
+
+`MessageAddr` is a composite key derived from the caller's entity address and a hash of the topic name:
+```rust
+pub struct MessageAddr {
+    /// The entity addr.
+    entity_addr: AddressableEntityHash,
+    /// A 32 byte BLAKE2b hash of the name of the message topic.
+    topic_hash: TopicNameHash,
+    /// The message index. If the message index is `None`, the `MessageAddr` points to the topic control record.
+    message_index: Option<u32>,
+}
+```
+
+There are 2 types of stored values that are written to global state under this key.
+
+The control record for the message topic:
+```rust
+pub struct MessageTopicSummary {
+    /// Number of messages in this topic.
+    pub(crate) message_count: u32,
+    /// Block timestamp in which these messages were emitted.
+    pub(crate) blocktime: BlockTime,
+}
+```
+
+
+Checksum of the message payload that was emitted:
+```rust
+pub struct MessageChecksum(
+    pub  [u8; 32], //a 32 byte BLAKE2b hash of the serialized `MessagePayload` that was emitted.
+);
+```
+
+When a new message is emitted on a specified topic the `message_count` in the `MessageTopicSummary` record is incremented.
+Messages have ephemeral nature by design. In order to not increase the size of global state in an uncontrolled manner, the message count is reset to zero with every newly emitted message that belongs to a new block and the block time is updated as well. The old message checksum entries are pruned from global state.
+Of-chain applications can still query the old message checksums if required by using the state root hash of the block where the messages were emitted.
+
+### Node changes
+
+The Execution engine provides a result to the node in the form of an internal `ExecutionResult` data structure. In order to pass the message along to the node we need to extend the `ExecutionResults` to include the messages that were emitted during execution on both success and failure of the deploy.
+The node will take the messages coming off the execution results in the `contract_runtime` component and will send them through the SSE hook as part of the `TransactionProcessed` SSE event.
+The messages are extracted and only sent out the SSE event. The external facing `ExecutionResults` type from `casper-types` will not contain the messages. This is because execution results get stored in node storage. The messages have an ephemeral nature and should not be stored.
+
+The `TransactionProcessed` SSE event will contain a `messages` field that will have an array of `Message` objects:
+
+```rust
+pub struct Message {
+    /// The identity of the entity that produced the message.
+    entity_addr: AddressableEntityHash,
+    /// The payload of the message.
+    message: MessagePayload,
+    /// The name of the topic on which the message was emitted on.
```

Comment:
```suggestion
    /// The name of the topic on which the message was emitted.
```

Reply:
Fixed in: https://github.com/casper-network/ceps/pull/88/commits/5ed76edf97bdf5e31cada98ce01522dad3e109d5